From 2b03ca5b8d326e182bb49bd8e5d28f7770c8fa6b Mon Sep 17 00:00:00 2001
From: Li Nan <linan122@huawei.com>
Date: Wed, 2 Nov 2022 15:15:23 +0800
Subject: [PATCH 161/595] [Huawei] iocost: add config option for iocost in
 cgroup v1

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187921

--------------------------------

Since 88c4110e635b ("cgroup: Support iocost for cgroup v1"), iocost is
enabled in cgroup v1. Provide BLK_CGROUP_LEGACY_IOCOST to control this
function.

Fixes: 88c4110e635b ("cgroup: Support iocost for cgroup v1")
Signed-off-by: Li Nan <linan122@huawei.com>
---
 block/Kconfig      | 8 ++++++++
 block/blk-iocost.c | 4 ++++
 2 files changed, 12 insertions(+)

diff --git a/block/Kconfig b/block/Kconfig
index e213002d5e73..123808a5bf90 100644
--- a/block/Kconfig
+++ b/block/Kconfig
@@ -155,6 +155,14 @@ config BLK_CGROUP_IOCOST
 	distributes IO capacity between different groups based on
 	their share of the overall weight distribution.
 
+config BLK_CGROUP_LEGACY_IOCOST
+	bool "Enable iocost in cgroup v1"
+	default n
+	depends on BLK_CGROUP_IOCOST=y
+	help
+	Enabling this option enables the .cost.weight interface for cost
+	model based proportional IO control when cgroup is version 1.
+
 config BLK_WBT_MQ
 	bool "Multiqueue writeback throttling"
 	default y
diff --git a/block/blk-iocost.c b/block/blk-iocost.c
index 9207850d3056..0a2ab2cd4fcf 100644
--- a/block/blk-iocost.c
+++ b/block/blk-iocost.c
@@ -3397,6 +3397,7 @@ static ssize_t ioc_cost_model_write(struct kernfs_open_file *of, char *input,
 	return ret;
 }
 
+#ifdef CONFIG_BLK_CGROUP_LEGACY_IOCOST
 static struct cftype ioc_legacy_files[] = {
 	{
 		.name = "cost.weight",
@@ -3418,6 +3419,7 @@ static struct cftype ioc_legacy_files[] = {
 	},
 	{}
 };
+#endif
 
 static struct cftype ioc_files[] = {
 	{
@@ -3443,7 +3445,9 @@ static struct cftype ioc_files[] = {
 
 static struct blkcg_policy blkcg_policy_iocost = {
 	.dfl_cftypes	= ioc_files,
+#ifdef CONFIG_BLK_CGROUP_LEGACY_IOCOST
 	.legacy_cftypes = ioc_legacy_files,
+#endif
 	.cpd_alloc_fn	= ioc_cpd_alloc,
 	.cpd_free_fn	= ioc_cpd_free,
 	.pd_alloc_fn	= ioc_pd_alloc,
-- 
2.17.1

